### ISSUE TITLE
Thread.terminate() called, yet mainprocess keeps hanging around.

### ISSUE THREAD

**Issue by philkunz** [andywer/threads.js#153]
The main process keeps hanging around when terminating all spawned workers With static function `Thread.terminate`

**Comment by philkunz** [comment]
Hm, this is really hard to circumvent for proper lifecycles.

**Comment by andywer** [comment]
I am a bit confused. The title says "terminate threads", but the description is about "The main process keeps hanging around".

Is this issue about `Thread.terminate()` not terminating workers (then please provide the node version you are using if running in node) or is it somehow related to the lifecycle of the master process?

**Comment by philkunz** [comment]
Yeah maybe the title is false :) 

Look here: https://gitlab.com/pushrocks/smartproxy/blob/master/ts/smartproxy.classes.smartproxy.ts

I'm trying to stop all workers, and I don't now why the main process keeps hanging around, as there should be nothing on the event loop scheduled after that, or anything that would keep the main process running.

**Comment by philkunz** [comment]
Basically I'm trying to build a traefik style high performance reverse proxy with ssl and websocket support in nodejs to replace nginx. https://www.npmjs.com/package/http-proxy-middleware is ridiculously slow since it a) doesn't keep around open tcp sockets after a request and b) since it is not multi threaded. However smartproxy should be stoppable in some decent way.

**Comment by andywer** [comment]
Can you try a little quick-fix? You can try adding `worker.unref()` to the `terminate` function in `createTerminator()` in `dist/master/spawn.js`.

Let me know if that works :)

**Comment by philkunz** [comment]
I get worker.unref is not a function, I'm using the tiny-worker fallback though.

I can see that https://nodejs.org/api/worker_threads.html#worker_threads_worker_unref should exists in the nodejs 12.x.x versions. However, due to your leakage package (and its dependency node-memwatch) only working on 10.x I'm using 10.x across my complete stack. 😅

**Comment by philkunz** [comment]
Ok, everything is closing as expected (even without worker.unref) in nodejs v12. So I just disabled leakage for now and switched to v12 with my stack. Thank you for your time though. Should I close this issue?

**Comment by andywer** [comment]
Cool. Yeah, let’s close it 👍

**Comment by keverw** [comment]
Yeah I ran into the same thing using the tiny-worker.

"Jest did not exit one second after the test run has completed."

I decided to just upgrade to Node 12 from 10 since the LTS will be out soon and it's a new project but the goal was to stick to the LTS for stability and in the future when multiple people working on it all can stick with the same stable version. Also writing unit tests so hopefully should catch things when upgrading anyways.

So reran the test and it worked perfectly as expected on V12.  So sounds like after you do `await Thread.terminate()` it is still doing somthing in the background.

Kinda annoying it installs `tiny-worker` even though not being used... but I'm also using yarn instead of npm too, so not sure if it does the same under NPM.

**Comment by andywer** [comment]
@keverw Can you try <https://github.com/andywer/threads.js/issues/153#issuecomment-524027080> for node 12?

For node 10 I just had an idea: Can you try adding this line to your worker?

```js
process.on("SIGINT", () => process.exit())
```

Let us know what happens 😉

**Comment by keverw** [comment]
On Line 92 after "yield worker.terminate();" added "yield worker.unref();" but didn't seem to make a difference but there's a chance I did something wrong since modifying the compiled code directly.

```
function createTerminator(worker) {
    const [termination, resolver] = promise_1.createPromiseWithResolver();
    const terminate = () => __awaiter(this, void 0, void 0, function* () {
        debugThreadUtils("Terminating worker");
        // Newer versions of worker_threads workers return a promise
        yield worker.terminate();
        yield worker.unref();
        resolver();
    });
    return { terminate, termination };
}
```

Haven't tried that idea with Node 10, but I could switch back and see. I probably should start using NVM...

**Comment by andywer** [comment]
The code sample looks right.

> I probably should start using NVM...

You should! 😉

> Was thinking maybe even though it terminates, maybe there's a slight delay instead of being instant even though I await it.

Yeah, I doubt it will terminate instantly. However, the process should then still exit once the workers have been terminated... I guess something else still keeps the process alive.

**Comment by keverw** [comment]
Yeah it seems if I run it without the test, it seems the process exits after a few seconds... So wonder if it's just the way Jest expects things to be, that it terminates instantly after the test.. Was thinking after the tests were done just waiting extra time would clear up the problem but I guess not... Not sure if i should worry about it much and just pass in that flag `--detectOpenHandles` it doesn't give the warning in yellow and closes after waiting but I'd think since it's called detect open handles it would at least tell me what was causing it.

**Comment by andywer** [comment]
Memo to myself: Let's try to debug using <https://www.npmjs.com/package/wtfnode>.

**Comment by keverw** [comment]
Hmm I added wtfnode to my test and see 
`      - (10000 ~ 10 s) (anonymous) @ node_modules/threads/dist/master/spawn.js:68` which seems related to observables somewhere in the code.

```
afterAll(async function (done) {
        wtf.dump();

        // to get around: Jest did not exit one second after the test run has completed.
        jest.setTimeout(20000);
        // await time.sleep(5000);

        setTimeout(function () {
            done();
        }, 15000);

    });
```

Seemed to clear it without the `--detectOpenHandles` flag... So I guess my theory was probably right then.

**Comment by andywer** [comment]
@keverw Which version of `threads` are you running?

`dist/master/spawn.js:68` is the line `const workerEvent = {` for me which doesn't seem right (it's this line in both `1.0.0-beta.3` and the latest `1.0.0-beta.4`).

**Comment by andywer** [comment]
@keverw  I think I fixed it! Can you try monkey-patching your `dist/master/spawn.js`?

Replace the `withTimeout()` function implementation (line 32ff):

```js
function withTimeout(promise, timeoutInMs, errorMessage) {
    return __awaiter(this, void 0, void 0, function* () {
        let timeoutHandle;
        const timeout = new Promise((resolve, reject) => {
            timeoutHandle = setTimeout(() => reject(Error(errorMessage)), timeoutInMs);
        });
        const result = yield Promise.race([
            promise,
            timeout
        ]);
        clearTimeout(timeoutHandle);
        return result;
    });
}
```

**Comment by andywer** [comment]
Thanks for the nice words, I appreciate it! Still waiting to be whitelisted for the new GitHub Sponsors feature. Not sure if it's gonna make a difference, though.

It's definitely not easy to monetize open source software without making it inaccessible to the hobby coders and small projects. Donations are good, but the fact that big commercial users have to opt-in to donating is quite a design flaw.

If anyone has a clever monetization strategy for a project like this to support its development in a sustainable way, don't hesitate to put it here in the issues 💸😄

**Comment by andywer** [comment]
Published as [`v1.0.0-beta.5`](https://github.com/andywer/threads.js/releases/tag/v1.0.0-beta.5) 🚀
